oauthproxy.go

```diff
@@ -1132,13 +1137,18 @@ func (p *OAuthProxy) getAuthenticatedSession(rw http.ResponseWriter, req *http.R
 // fails the linter. Remove the nolint when functionality expands.
 //
 //nolint:gosimple
-func authOnlyAuthorize(req *http.Request, s *sessionsapi.SessionState) bool {
+func (p *OAuthProxy) authOnlyAuthorize(req *http.Request, s *sessionsapi.SessionState) bool {
 	// Allow secondary group restrictions based on the `allowed_groups`
 	// querystring parameter
 	if !checkAllowedGroups(req, s) {
 		return false
 	}
 
+	// Transfer authorization to Open Policy Agent
+	if !p.checkOpenPolicyAgent(req, s) {
```

Comment:
This is an interesting addition!

I think as we look to add extendable `authorizers`, it might be more clearly organized if we get an `authrorizer` interface and make a package area to house these implementations so we aren't making the already massive `oauthproxy.go` even larger.

Reply:
For more context - @JoelSpeed and I have been brainstorming what an authorization interface would look like in the main execution path (inline proxy, through an `Authorize` method).

Extendable to internal rules, OPA (central server), [Keto](https://github.com/ory/keto) (central server), [Oso](https://github.com/osohq/oso) (sidecar).

End goal of having more flexible authorization and replacing the existing global IP, path or provider groups logic with fine-grained policies.

Reply:
I did also have a play with integrating the OPA Go library into the proxy, it didn't go so well and got incredibly complex! I think we need to make sure that we aren't extending the request duration too much by forking out to a separate web service

Reply:
I think an authrorizer interface independent of the oauth2 provider make sense.

Reply:
I'm not sure is the OPA GO library is the correct solution because the different solutions ([REST, Go library and WASM](https://www.openpolicyagent.org/docs/latest/integration/)) have different update behaviors.

Nginx supports a cache for the auth_request to reduce the impact to the requests.